### PR TITLE
fix: improve error message when persistTransaction times out. Increase default timeout to 130 seconds using Exponential Backoff

### PR DESCRIPTION
**Pull request by MantisClone**
## Description of the changes

* Update the default getConfirmation timeout to use exponential backoff by default:  starting at 1s, doubling after each retry, up to a maximum of 64s and max 7 retries with an initial 3s defer delay, yielding a total of 8 calls and total timeout of 130s.
* Update timeout error message to more adequately reassure the App Builder not to call persistTransaction again, and to more fully explain what they can do instead (poll `getConfirmedTransaction()` and adjust their `httpConfig` settings so timeouts are less likely).
* Move `getConfirmedTransaction` logic to a new public function in the request-client.js package.
* Adjust retry logic to make `exponentialBackoffDelay` more intuitive to configure
  * Now, 1000 means retry after 1, 2, 4, 8... seconds.
  * Previously, 1000 meant retry after 2, 4, 8, 16... seconds.

## TODO

- [x] Make sure the new error message doesn't break the timeout dialog in the Invoicing Template added in https://github.com/RequestNetwork/web-components/pull/99
  * :x: These changes break Request Invoicing so the Create Request Form Web Component needs to be updated if/when it uses the latest RN SDK libraries (after this PR is merged and released). https://github.com/RequestNetwork/web-components/blob/5eb3818f1538204d0fb7af14bedf45e8fd5a870d/packages/create-invoice-form/src/lib/create-invoice-form.svelte#L154C9-L154C9
    ```
            if (error.message.includes("Transactioon confirmation not received")) {
    ```

## Summary by CodeRabbit

- **New Features**
	- Added a "V1 packages" section in the README for clarity on deprecated packages.
	- Introduced a new method, `getConfirmedTransaction`, for retrieving confirmed transactions.

- **Bug Fixes**
	- Enhanced error messages for transaction confirmation failures to guide app builders.

- **Documentation**
	- Expanded testing instructions in the README.
	- Improved clarity and structure in the README content.

- **Tests**
	- Updated tests for `persistTransaction` to reflect new error handling.
	- Revised tests for the `retry` function to align with the new exponential backoff strategy.

### CODE REVIEW

**Comment by MantisClone** [comment]
I'm reducing the scope by removing the exponential backoff. Instead, set the default timeout to 2 minutes, up from 30 seconds, to match the GRAPH_IPFS_TIMEOUT on the storage subgraph.

For posterity, here's the proposed exponential backoff change:

Enable exponential backoff by default. 

Proposed:
* getConfirmationDeferDelay: 3000
* getConfirmationRetryDelay: 0
* getConfirmationExponentialBackoffDelay: 1000
* getConfirmationMaxExponentialBackoffDelay: 33000

This should make the confirmation polling happen like:
1. check after 3 seconds (defer delay)
2. check after 4 seconds (start exponential backoff)
3. check after 5 seconds
4. check after 7 seconds
5. check after 11 seconds
6. check after 19 seconds
7. check after 35 seconds
8. Give up after 36 seconds. (defer + max exponential backoff) // Need a test to verify that it actually gives up after 36 seconds

File: packages/request-client.js/src/http-data-access.ts [hunk 1]
```diff
@@ -109,33 +107,20 @@ export default class HttpDataAccess implements DataAccessTypes.IDataAccess {
...
-              transactionHash,
-            },
-            {
-              maxRetries: this.httpConfig.getConfirmationMaxRetry,
-              retryDelay: this.httpConfig.getConfirmationRetryDelay,
-              exponentialBackoffDelay: this.httpConfig.getConfirmationExponentialBackoffDelay,
-              maxExponentialBackoffDelay: this.httpConfig.getConfirmationMaxExponentialBackoffDelay,
-            },
-          );
+        const confirmedData = await this.getConfirmedTransaction(transactionData);
```
Comment (MantisClone) [inline packages/request-client.js/src/http-data-access.ts:110]: I moved this logic into a new `getConfirmedTransaction()` function

File: packages/request-client.js/test/http-data-access.test.ts [hunk 2]
```diff
@@ -29,22 +29,14 @@ describe('HttpDataAccess', () => {
...
-          maxExponentialBackoffDelay: 30000ms`),
-      );
+      void httpDataAccess.persistTransaction({}, '', []).then((returnPersistTransaction) => {
+        returnPersistTransaction.on('error', (e: any) => {
+          expect(e.message).toBe(
+            `The Request Network SDK timed-out while polling the Request Node to confirm that the Request was persisted successfully. It is likely that the persisted Request will be confirmed eventually. App Builders are discouraged from calling persistTransaction() a second time, which would create a duplicate Request. Instead, App Builders are recommended to catch this error and continue polling the Request Node using getConfirmedTransaction() or by calling the /getConfirmedTransaction endpoint.  To avoid timeouts in the future, try adjusting the httpConfig values when instantiating the RequestNetwork object: getConfirmationDeferDelay: 0ms, getConfirmationMaxRetries: 0, getConfirmationRetryDelay: 1000ms, getConfirmationExponentialBackoffDelay: 0ms, getConfirmationMaxExponentialBackoffDelay: 30000ms`,
+          );
+          done();
+        });
+      });
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/request-client.js/test/http-data-access.test.ts]: Agreed :+1: Fixed :white_check_mark:

File: README.md [hunk 3]
```diff
@@ -83,6 +83,12 @@ Test all the packages in the monorepo.
 yarn run test
 ```
 
+Test a specific package.
+
+```bash
+yarn workspace @requestnetwork/package-name test
+```
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply README.md]: Good idea :+1: Done :white_check_mark:

File: packages/utils/test/retry.test.ts [hunk 4]
```diff
@@ -132,37 +132,96 @@ describe('Retry', () => {
     });
 
     retry(throwFn, {
-      exponentialBackoffDelay: 1000,
-      maxExponentialBackoffDelay: 7000,
+      maxRetries: 30,
+      retryDelay: 0,
+      exponentialBackoffDelay: 1000, // 1s
+      maxExponentialBackoffDelay: 122000, // 122s
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/utils/test/retry.test.ts:138]: Good catch :+1: Fixed :white_check_mark:

File: packages/utils/test/retry.test.ts [hunk 5]
```diff
@@ -132,37 +132,96 @@
...
+
+    // Call 9th time after 120s (247000ms total elapsed)
+    jest.advanceTimersByTime(119999);
+    await Promise.resolve();
+    expect(throwFn).toHaveBeenCalledTimes(8);
+    jest.advanceTimersByTime(120000);
+    await Promise.resolve();
+    expect(throwFn).toHaveBeenCalledTimes(9);
+
+    expect(jest.getRealSystemTime()).toBe(247000);
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/utils/test/retry.test.ts:224]: Makes sense :+1:  Done :white_check_mark:

File: README.md [hunk 6]
```diff
@@ -83,6 +83,12 @@ Test all the packages in the monorepo.
 yarn run test
 ```
 
+Test a specific package.
+
+```bash
+yarn workspace @requestnetwork/request-client.js test
+```
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply README.md:90]: Okay :+1: Done :white_check_mark:

File: packages/request-client.js/src/http-data-access.ts [hunk 7]
```diff
@@ -109,33 +107,15 @@ export default class HttpDataAccess implements DataAccessTypes.IDataAccess {
...
-          maxExponentialBackoffDelay: ${this.httpConfig.getConfirmationMaxExponentialBackoffDelay}ms`,
+            `The Request Network SDK timed-out while polling the Request Node to confirm that the Request was persisted successfully. It is likely that the persisted Request will be confirmed eventually. App Builders are discouraged from calling persistTransaction() a second time, which would create a duplicate Request. Instead, App Builders are recommended to catch this error and continue polling the Request Node using getConfirmedTransaction() or by calling the /getConfirmedTransaction endpoint.  To avoid timeouts in the future, try adjusting the httpConfig values when instantiating the RequestNetwork object: getConfirmationDeferDelay: ${this.httpConfig.getConfirmationDeferDelay}ms, getConfirmationMaxRetries: ${this.httpConfig.getConfirmationMaxRetry}, getConfirmationRetryDelay: ${this.httpConfig.getConfirmationRetryDelay}ms, getConfirmationExponentialBackoffDelay: ${this.httpConfig.getConfirmationExponentialBackoffDelay}ms, getConfirmationMaxExponentialBackoffDelay: ${this.httpConfig.getConfirmationMaxExponentialBackoffDelay}ms`,
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/request-client.js/src/http-data-access.ts:118]: Good catch :+1: Done :white_check_mark:

File: packages/utils/test/retry.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/utils/test/retry.test.ts:224]: Okay :+1: Fixed :white_check_mark:

File: packages/request-client.js/test/http-data-access.test.ts [hunk 8]
```diff
@@ -29,22 +29,20 @@ describe('HttpDataAccess', () => {
...
+          errorEmitted = true;
+          expect(e.message).toBe(
+            `The Request Network SDK timed-out while polling the Request Node to confirm that the Request was persisted successfully. It is likely that the persisted Request will be confirmed eventually. App Builders are discouraged from calling persistTransaction() a second time, which would create a duplicate Request. Instead, App Builders are recommended to catch this error and continue polling the Request Node using getConfirmedTransaction() or by calling the /getConfirmedTransaction endpoint.  To avoid timeouts in the future, try adjusting the httpConfig values when instantiating the RequestNetwork object: getConfirmationDeferDelay: 0ms, getConfirmationMaxRetries: 0, getConfirmationRetryDelay: 1000ms, getConfirmationExponentialBackoffDelay: 0ms, getConfirmationMaxExponentialBackoffDelay: 30000ms`,
+          );
+        });
+
+        setTimeout(() => {
+          expect(errorEmitted).toBe(true);
+          done();
+        }, 1000);
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/request-client.js/test/http-data-access.test.ts]: Okay :+1: Fixed :white_check_mark:

File: packages/utils/test/retry.test.ts [hunk 9]
```diff
@@ -132,37 +132,88 @@ describe('Retry', () => {
...
     retry(throwFn, {
-      exponentialBackoffDelay: 1000,
-      maxExponentialBackoffDelay: 7000,
-    })();
+      retryDelay: 0,
+      // Exponential backoff starting at 1s, doubling each time, up to a maximum of 64s and max 8 retries, yielding a total timeout of 127s
+      maxRetries: 7,
+      exponentialBackoffDelay: 1000, // 1s
+      maxExponentialBackoffDelay: 64000, // 64s
+    })().catch(() => {
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/utils/test/retry.test.ts:140]: Good point :thumbs_up: Fixed :white_check_mark:

File: packages/request-client.js/test/http-data-access.test.ts [hunk 10]
```diff
@@ -29,22 +29,15 @@ describe('HttpDataAccess', () => {
...
-      );
+      const returnPersistTransaction = await httpDataAccess.persistTransaction({}, '', []);
+      await new Promise<void>((resolve) => {
+        returnPersistTransaction.on('error', (e: any) => {
+          expect(e.message).toBe(
+            `The Request Network SDK timed-out while polling the Request Node to confirm that the Request was persisted successfully. It is likely that the persisted Request will be confirmed eventually. App Builders are discouraged from calling persistTransaction() a second time, which would create a duplicate Request. Instead, App Builders are recommended to catch this error and continue polling the Request Node using getConfirmedTransaction() or by calling the /getConfirmedTransaction endpoint.  To avoid timeouts in the future, try adjusting the httpConfig values when instantiating the RequestNetwork object: getConfirmationDeferDelay: 0ms, getConfirmationMaxRetries: 0, getConfirmationRetryDelay: 1000ms, getConfirmationExponentialBackoffDelay: 0ms, getConfirmationMaxExponentialBackoffDelay: 30000ms`,
+          );
+          resolve();
+        });
+      });
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/request-client.js/test/http-data-access.test.ts]: Sure :+1: Done :white_check_mark:

File: packages/request-client.js/src/http-data-access.ts [hunk 11]
```diff
@@ -144,6 +124,29 @@ export default class HttpDataAccess implements DataAccessTypes.IDataAccess {
...
+      },
+      {
+        maxRetries: this.httpConfig.getConfirmationMaxRetry,
+        retryDelay: this.httpConfig.getConfirmationRetryDelay,
+        exponentialBackoffDelay: this.httpConfig.getConfirmationExponentialBackoffDelay,
+        maxExponentialBackoffDelay: this.httpConfig.getConfirmationMaxExponentialBackoffDelay,
+      },
+    );
+  }
+
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/request-client.js/src/http-data-access.ts:149]: In this case, I don't think we need to throw an error if it fails.

File: packages/request-client.js/src/http-config-defaults.ts [hunk 12]
```diff
@@ -6,10 +6,12 @@ const config: ClientTypes.IHttpDataAccessConfig = {
...
-  getConfirmationMaxRetry: 30,
-  getConfirmationRetryDelay: 1000,
-  getConfirmationExponentialBackoffDelay: 0,
-  getConfirmationMaxExponentialBackoffDelay: 30000,
+
+  // Exponential backoff starting at 1s, doubling after each retry, up to a maximum of 64s and max 7 retries with an initial 3s defer delay, yielding a total of 8 calls and total timeout of 130s
+  getConfirmationMaxRetry: 7,
+  getConfirmationRetryDelay: 0,
+  getConfirmationExponentialBackoffDelay: 1000,
+  getConfirmationMaxExponentialBackoffDelay: 64000,
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/request-client.js/src/http-config-defaults.ts:14]: We won't add additional logging or metrics at this time.

File: packages/utils/test/retry.test.ts [hunk 13]
```diff
@@ -132,37 +132,88 @@ describe('Retry', () => {
...
+    jest.advanceTimersByTime(1);
+    await Promise.resolve();
+    expect(throwFn).toHaveBeenCalledTimes(8);
+    expect(Date.now()).toBe(127000);
+
+    // No further retries
+    jest.advanceTimersByTime(1000000000);
+    await Promise.resolve();
+    expect(throwFn).toHaveBeenCalledTimes(8);
+    expect(Date.now()).toBe(1000127000);
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/utils/test/retry.test.ts:216]: While I like the brevity of the proposed refactor, I quite like how the current test unrolls the loop and includes comments showing, at a glance, the number of calls and total elapsed time for any given number of retries. So I won't refactor in this case.

File: packages/utils/src/retry.ts [hunk 14]
```diff
@@ -61,7 +61,7 @@ const retry = <TParams extends unknown[], TReturn>(
             setTimeout(
               resolve,
               retryDelay +
-                Math.min(maxExponentialBackoffDelay, exponentialBackoffDelay * 2 ** retry),
+                Math.min(maxExponentialBackoffDelay, (exponentialBackoffDelay / 2) * 2 ** retry),
```
Comment (MantisClone) [inline packages/utils/src/retry.ts:64]: This change adjusts the retry logic to make `exponentialBackoffDelay` more intuitive to configure
* Now, 1000 means retry after 1, 2, 4, 8... seconds.
* Previously, 1000 meant retry after 2, 4, 8, 16... seconds.

File: packages/request-client.js/test/http-data-access.test.ts [hunk 15]
```diff
@@ -29,22 +29,18 @@ describe('HttpDataAccess', () => {
...
-          exponentialBackoffDelay: 0ms,
-          maxExponentialBackoffDelay: 30000ms`),
-      );
+      const returnPersistTransaction = await httpDataAccess.persistTransaction({}, '', []);
+      await Promise.race([
+        new Promise<void>((resolve) => {
+          returnPersistTransaction.on('error', (e: any) => {
+            expect(e.message).toBe(
+              'Timeout while confirming the Request was persisted. It is likely that the Request will be confirmed eventually. Recommend catching this error and using getConfirmedTransaction() to continue polling for confirmation. Consider adjusting the httpConfig settings on the RequestNetwork object to avoid future timeouts. Discourage calling persistTransaction() again as this would create a duplicate Request.',
+            );
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/request-client.js/test/http-data-access.test.ts:38]: Unnecessary.

File: packages/request-client.js/test/http-data-access.test.ts [hunk 16]
```diff
@@ -29,22 +29,18 @@ describe('HttpDataAccess', () => {
...
-          deferDelay: 0ms,
-          maxRetries: 0,
-          retryDelay: 1000ms,
-          exponentialBackoffDelay: 0ms,
-          maxExponentialBackoffDelay: 30000ms`),
-      );
+      const returnPersistTransaction = await httpDataAccess.persistTransaction({}, '', []);
+      await Promise.race([
+        new Promise<void>((resolve) => {
+          returnPersistTransaction.on('error', (e: any) => {
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/request-client.js/test/http-data-access.test.ts:35]: Can't due to error:

```
Argument of type '(e: Error) => void' is not assignable to parameter of type '(error: unknown) => void'.
  Types of parameters 'e' and 'error' are incompatible.
    Type 'unknown' is not assignable to type 'Error'.ts(2345)
```

File: packages/request-client.js/src/http-data-access.ts [hunk 17]
```diff
@@ -109,33 +107,15 @@ export default class HttpDataAccess implements DataAccessTypes.IDataAccess {
...
         if (e.status === 404) {
           error = new Error(
-            `Transaction confirmation not received. Try polling
-          getTransactionsByChannelId() until the transaction is confirmed.
-          deferDelay: ${this.httpConfig.getConfirmationDeferDelay}ms,
-          maxRetries: ${this.httpConfig.getConfirmationMaxRetry},
-          retryDelay: ${this.httpConfig.getConfirmationRetryDelay}ms,
-          exponentialBackoffDelay: ${this.httpConfig.getConfirmationExponentialBackoffDelay}ms,
-          maxExponentialBackoffDelay: ${this.httpConfig.getConfirmationMaxExponentialBackoffDelay}ms`,
+            `Timeout while confirming the Request was persisted. It is likely that the Request will be confirmed eventually. Recommend catching this error and using getConfirmedTransaction() to continue polling for confirmation. Consider adjusting the httpConfig settings on the RequestNetwork object to avoid future timeouts. Discourage calling persistTransaction() again as this would create a duplicate Request.`,
```
[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/request-client.js/src/http-data-access.ts:118]: Nice one :+1:  Done :white_check_mark:

[the remark this answers is not in the document]
Comment (MantisClone) [reply packages/request-client.js/src/http-data-access.ts:118]: Okay :+1:  Done :white_check_mark:

File: packages/request-client.js/src/http-data-access.ts [hunk 18]
```diff
@@ -109,33 +107,15 @@ export default class HttpDataAccess implements DataAccessTypes.IDataAccess {
...
-            },
-          );
+        const confirmedData = await this.getConfirmedTransaction(transactionData);
         // when found, emit the event 'confirmed'
         result.emit('confirmed', confirmedData);
       })
       .catch((e) => {
         let error: Error = e;
-        if (e.status === 404) {
+        if (e && 'status' in e && e.status === 404) {
```
Comment (MantisClone) [inline packages/request-client.js/src/http-data-access.ts:116]: This was a coderabbit suggestion.

**Review by rodrigopavezi — APPROVED** [review]
great job 💯

**Review by aimensahnoun — APPROVED** [review]

**Pull request merged**